src/sentry/web/frontend/accept_organization_invite.py

```diff
@@ -108,52 +100,129 @@ def handle(self, request, member_id, token):
-                    (organization.name.encode('utf-8'), )
-                )
-
-                om.delete()
-            else:
-                om.user = request.user
-                om.email = None
-                om.save()
-
-                self.create_audit_entry(
-                    request,
-                    organization=organization,
-                    target_object=om.id,
-                    target_user=request.user,
-                    event=AuditLogEntryEvent.MEMBER_ACCEPT,
-                    data=om.get_audit_log_data(),
-                )
-
-                messages.add_message(
-                    request, messages.SUCCESS,
-                    _('You have been added to the %r organization.') %
-                    (organization.name.encode('utf-8'), )
-                )
-
-                member_joined.send_robust(member=om, organization=organization, sender=self)
+            helper.accept_invite()
 
             request.session.pop('can_register', None)
             response = self.redirect(reverse('sentry-organization-home', args=[organization.slug]))
-
-            if PENDING_INVITE in request.COOKIES:
-                response.delete_cookie(PENDING_INVITE)
-            return response
+            return helper.remove_invite_cookie(response)
 
         context['form'] = form
-
         return self.respond('sentry/accept-organization-invite.html', context)
+
+
+class BaseInviteHelper(object):
+    def __init__(self, instance, request, member_id, token, logger=None):
+        self.request = request
+        self.instance = instance
+        self.member_id = member_id
+        self.token = token
+        self.logger = logger
+
+    def handle_success(self):
+        pass
+
+    def handle_member_already_exists(self):
+        pass
+
+    def get_organization_member(self):
+        self.om = OrganizationMember.objects.select_related('organization').get(pk=self.member_id)
+        return self.om
+
+    @property
+    def member_pending(self):
+        return self.om.is_pending
+
+    @property
+    def valid_token(self):
+        return constant_time_compare(self.om.token or self.om.legacy_token, self.token)
+
+    @property
+    def user_authenticated(self):
+        return self.request.user.is_authenticated()
+
+    @property
+    def needs_2fa(self):
+        org_requires_2fa = self.om.organization.flags.require_2fa.is_set
+        user_has_2fa = Authenticator.objects.user_has_2fa(self.request.user.id)
+        return org_requires_2fa and not user_has_2fa
+
+    @property
+    def member_already_exists(self):
+        return OrganizationMember.objects.filter(
+            organization=self.om.organization, user=self.request.user
+        ).exists()
+
+    def accept_invite(self):
+        om = self.om
+
+        if self.member_already_exists:
+            self.handle_member_already_exists()
+            om.delete()
+        else:
+            om.user = self.request.user
```

Comment:
What ensures that `om` is not None? Right now it looks like `accept_invite` could be called before `get_organization_member` resulting in a runtime error here. Should `get_organization_member` be called as part of the helper's constructor?